File: WordSquare.java
/**
 * @author Nishant M.
 * @date 26-June-2012
 * @problem http://en.wikipedia.org/wiki/Word_square
 *
 */
public class WordSquare{
	
	public static void main(String[] args) {
		
		/*
		 * Example Test.
		 *------------------------------------------------------
		 *	   5			4		  3			2
		 * H E A R T	 H E A R	H E A		H E
		 * E M B E R	 E M B E	E M B		E M
		 * A B U S E	 A B U S	A B U
		 * R E S I N	 R E S I	
		 * T R E N D
		 * 
		 */
		String[] inputStrings = {"HEART","EMBER","ABUSE","RESIN","TREND"};
		String[][] arr = build2D(inputStrings);
		int lengthofword= inputStrings[0].length();	//ASSUMES all words of same length. Does not verify.
		System.out.println("Input is a Word Sqaure ?  "+isSquare(arr,lengthofword));
	}

	private static boolean isSquare(String[][] arr, int lengthofword) {
		if(lengthofword>2){
			if(isSquare(arr, lengthofword-1)){
				return getWords(arr, lengthofword);
			}
			else
				return false;
		}else{
			return getWords(arr,lengthofword);
		}
	}

	private static boolean getWords(String[][] arr, int lengthofword) {
		String one = "",two = "";
		for (int i = 0; i <lengthofword ; i++) {
			one+=arr[i][lengthofword-1];
			two+=arr[lengthofword-1][i];
		}
		if(one.equalsIgnoreCase(two))
			return true;
		else
			return false;
	}

	private static String[][] build2D(String[] in) {
		String[][] arr = new String[in.length][in.length];
		for (int i = 0; i < in.length; i++) {
			for (int j = 0; j < in.length; j++) {
				arr[i][j]=""+in[i].charAt(j);
			}
		}
		return arr;
	}
}

File: WordSquare.rb
# @author Nishant M.
# @date 05-February-2013
# @problem http://en.wikipedia.org/wiki/Word_square

class WordSquare 

	def process
		inputStrings = %w{HEART EMBER ABUSE RESIN TREND}
		twoD = Array.new()
		inputStrings.each {|a| twoD << a.chars.to_a }
		puts "Input is a Word Square ? " + isSquare(twoD,inputStrings[1].length)
 	end

 	def isSquare(twod,len)
 		if len > 2
 			if isSquare(twod,len-1)
 				getWords(twod,len)
 			else
 				"false"
 			end
 		else
 			getWords(twod,len)	
 		end
 	end

 	def getWords(a,len)
 		one = Array.new
 		two = Array.new
 		len.times do |i|
 			one << a[i][len-1]
 			two << a[len-1][i]
 		end
 		if one == two
 			"true"
 		else
 			"false"
 		end
 	end
end

w = WordSquare.new
w.process